main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer1.scala

```diff
@@ -0,0 +1,580 @@
+/*
+ * Copyright 2022 Anna Krogh, Patrick Lundvig, Christian Bonde
+ * 2024 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.MonoAst
+import ca.uwaterloo.flix.language.ast.OccurrenceAst1.Occur.*
+import ca.uwaterloo.flix.language.ast.OccurrenceAst1.{DefContext, Occur}
+import ca.uwaterloo.flix.language.ast.Symbol.{DefnSym, VarSym}
+import ca.uwaterloo.flix.language.ast.{AtomicOp, OccurrenceAst1, Symbol}
+import ca.uwaterloo.flix.util.ParOps
+
+/**
+  * The occurrence analyzer collects information on variable and function usage and calculates the weight of the expressions
```

Comment:
maybe some intuition on what the weight is

Reply:
Yeah that is still todo